<a href="https://colab.research.google.com/github/vishva2410/Deep-Learning-Projects-Part-A-/blob/main/finacial_rating.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

In [7]:
df = pd.read_csv("financial_rating_result.csv")
df.head(4)

Income  Age  Dependents  Occupation  City_Tier          Rent  \
0   44637.249636   49           0           2          0  13391.174891   
1   26858.596592   34           2           1          1   5371.719318   
2   50367.605084   35           1           3          2   7555.140763   
3  101455.600247   21           0           2          2  15218.340037   

   Loan_Repayment    Insurance     Groceries    Transport  ...  \
0        0.000000  2206.490129   6658.768341  2636.970696  ...   
1        0.000000   869.522617   2818.444460  1543.018778  ...   
2     4612.103386  2201.800050   6313.222081  3221.396403  ...   
3     6809.441427  4889.418087  14690.149363  7106.130005  ...   

   Disposable_Income  Potential_Savings_Groceries  \
0       11265.627707                  1685.696222   
1        9676.818733                   540.306561   
2       13891.450624                  1466.073984   
3       31617.953615                  1875.932770   

   Potential_Savings_Transport  Potential_Savings_Eating_Out  \
0                   328.895281                    465.769172   
1                   119.347139                    141.866089   
2                   473.549752                    410.857129   
3                   762.020789                   1241.017448   

   Potential_Savings_Entertainment  Potential_Savings_Utilities  \
0                       195.151320                   678.292859   
1                       234.131168                   286.668408   
2                       459.965256                   488.383423   
3                       320.190594                  1389.815033   

   Potential_Savings_Healthcare  Potential_Savings_Education  \
0                     67.682471                     0.000000   
1                      6.603212                    56.306874   
2                      7.290892                   106.653597   
3                    193.502754                     0.000000   

   Potential_Savings_Miscellaneous  overall_score  
0                        85.735517           3.79  
1                        97.388606           4.36  
2                       138.542422           6.62  
3                       296.041183           5.03  

[4 rows x 28 columns]

In [9]:
X = df.drop(columns=['overall_score'])
y = df['overall_score']

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
sd = StandardScaler()
X_train_scaled = sd.fit_transform(X_train)
X_test_scaled = sd.transform(X_test)

In [12]:
# Train a linear regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

print("Model training complete.")

Model training complete.


R-squared score: 1.0000


# Task
Analyze the financial rating data from "/content/financial_rating_result.csv", build a regression model to predict the 'Overall Score', evaluate the model's accuracy using various metrics, and create a user interface for predicting the financial rating based on user input.

## Evaluate model

### Subtask:
Calculate additional regression metrics like Mean Squared Error (MSE) and Mean Absolute Error (MAE).


**Reasoning**:
Calculate and print MSE and MAE.



In [14]:
from sklearn.metrics import mean_squared_error, mean_absolute_error

y_pred = model.predict(X_test_scaled)

mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"Mean Absolute Error (MAE): {mae:.4f}")

Mean Squared Error (MSE): 0.0000
Mean Absolute Error (MAE): 0.0025


## Create prediction function

### Subtask:
Define a Python function that takes financial details as input, preprocesses them using the same scaler, and outputs the predicted financial rating score using the trained model.


**Reasoning**:
Define a function to predict the financial rating based on user input, preprocess the input using the pre-fitted scaler, and use the trained model for prediction.



In [15]:
import pandas as pd

def predict_financial_rating(income, age, dependents, occupation, city_tier, rent, loan_repayment, insurance, groceries, transport, entertainment, utilities, healthcare, education, miscellaneous):
  """
  Predicts the financial rating based on user input financial details.

  Args:
    income: Annual income.
    age: Age of the individual.
    dependents: Number of dependents.
    occupation: Occupation type (numerical representation).
    city_tier: City tier (numerical representation).
    rent: Monthly rent.
    loan_repayment: Monthly loan repayment.
    insurance: Monthly insurance cost.
    groceries: Monthly groceries cost.
    transport: Monthly transport cost.
    entertainment: Monthly entertainment cost.
    utilities: Monthly utilities cost.
    healthcare: Monthly healthcare cost.
    education: Monthly education cost.
    miscellaneous: Monthly miscellaneous expenses.

  Returns:
    Predicted financial rating score.
  """
  input_data = {
      'Income': [income],
      'Age': [age],
      'Dependents': [dependents],
      'Occupation': [occupation],
      'City_Tier': [city_tier],
      'Rent': [rent],
      'Loan_Repayment': [loan_repayment],
      'Insurance': [insurance],
      'Groceries': [groceries],
      'Transport': [transport],
      'Entertainment': [entertainment],
      'Utilities': [utilities],
      'Healthcare': [healthcare],
      'Education': [education],
      'Miscellaneous': [miscellaneous],
      'Disposable_Income': [income - rent - loan_repayment - insurance - groceries - transport - entertainment - utilities - healthcare - education - miscellaneous],
      'Potential_Savings_Rent': [0.15 * rent],
      'Potential_Savings_Loan_Repayment': [0.15 * loan_repayment],
      'Potential_Savings_Insurance': [0.15 * insurance],
      'Potential_Savings_Groceries': [0.15 * groceries],
      'Potential_Savings_Transport': [0.15 * transport],
      'Potential_Savings_Eating_Out': [0.15 * (0.2 * income)], # Assuming 20% of income is spent on eating out and 15% of that can be saved
      'Potential_Savings_Entertainment': [0.15 * entertainment],
      'Potential_Savings_Utilities': [0.15 * utilities],
      'Potential_Savings_Healthcare': [0.15 * healthcare],
      'Potential_Savings_Education': [0.15 * education],
      'Potential_Savings_Miscellaneous': [0.15 * miscellaneous]
  }
  input_df = pd.DataFrame(input_data)

  # Ensure column order matches training data
  input_df = input_df[X_train.columns]

  input_scaled = sd.transform(input_df)
  predicted_score = model.predict(input_scaled)
  return predicted_score[0]

print("Function 'predict_financial_rating' defined.")

Function 'predict_financial_rating' defined.


## Build user interface

### Subtask:
Create a simple interface using `ipywidgets` or a similar library in Colab that allows users to input their financial details and get a prediction using the function created in the previous step.


**Reasoning**:
Create a simple interface using ipywidgets for user input and prediction.



In [16]:
from ipywidgets import VBox, FloatText, IntText, Button, Output
from IPython.display import display

# Create input widgets
income_input = FloatText(description='Income:')
age_input = IntText(description='Age:')
dependents_input = IntText(description='Dependents:')
occupation_input = IntText(description='Occupation (0-3):') # Assuming occupation is encoded as 0, 1, 2, 3
city_tier_input = IntText(description='City Tier (0-2):') # Assuming city tier is encoded as 0, 1, 2
rent_input = FloatText(description='Rent:')
loan_repayment_input = FloatText(description='Loan Repayment:')
insurance_input = FloatText(description='Insurance:')
groceries_input = FloatText(description='Groceries:')
transport_input = FloatText(description='Transport:')
entertainment_input = FloatText(description='Entertainment:')
utilities_input = FloatText(description='Utilities:')
healthcare_input = FloatText(description='Healthcare:')
education_input = FloatText(description='Education:')
miscellaneous_input = FloatText(description='Miscellaneous:')


# Create a button for prediction
predict_button = Button(description='Predict Financial Rating')

# Create an output widget to display the result
output_widget = Output()

# Define the function to be executed when the button is clicked
def on_predict_button_clicked(b):
    with output_widget:
        output_widget.clear_output()
        try:
            # Get values from input widgets
            income = income_input.value
            age = age_input.value
            dependents = dependents_input.value
            occupation = occupation_input.value
            city_tier = city_tier_input.value
            rent = rent_input.value
            loan_repayment = loan_repayment_input.value
            insurance = insurance_input.value
            groceries = groceries_input.value
            transport = transport_input.value
            entertainment = entertainment_input.value
            utilities = utilities_input.value
            healthcare = healthcare_input.value
            education = education_input.value
            miscellaneous = miscellaneous_input.value


            # Call the prediction function
            predicted_score = predict_financial_rating(income, age, dependents, occupation, city_tier, rent, loan_repayment, insurance, groceries, transport, entertainment, utilities, healthcare, education, miscellaneous)

            # Display the prediction
            print(f"Predicted Financial Rating: {predicted_score:.2f}")
        except Exception as e:
            print(f"Error: {e}")

# Link the button's click event to the function
predict_button.on_click(on_predict_button_clicked)

# Arrange widgets in a vertical box
ui = VBox([
    income_input,
    age_input,
    dependents_input,
    occupation_input,
    city_tier_input,
    rent_input,
    loan_repayment_input,
    insurance_input,
    groceries_input,
    transport_input,
    entertainment_input,
    utilities_input,
    healthcare_input,
    education_input,
    miscellaneous_input,
    predict_button,
    output_widget
])

# Display the interface
display(ui)

## Summary:

### Data Analysis Key Findings

*   The trained regression model for predicting 'Overall Score' demonstrated high accuracy on the test data, with a Mean Squared Error (MSE) of approximately 0.0000 and a Mean Absolute Error (MAE) of approximately 0.0025.
*   A Python function `predict_financial_rating` was successfully created to predict the financial rating score based on user-provided financial details. This function preprocesses the input using a pre-fitted scaler and utilizes the trained regression model.
*   An interactive user interface was built using `ipywidgets` in Colab, allowing users to input their financial details and obtain a predicted financial rating.

### Insights or Next Steps

*   The model's high accuracy suggests it is performing very well on the provided dataset. It would be beneficial to test the model on new, unseen data to further validate its performance and generalizability.
*   The created user interface provides a practical way for users to interact with the prediction model. Further enhancements could include adding input validation and clearer descriptions for the input fields.
